Rework kvstore

fmorency opened the pull request:
MVP++

- Add account/event support
- Add query endpoint
- Rename `delete` to `disable`
- Add tests
- Port staging file to json5
- Fix Bazel
- Add kvstore e2e tests
- kvstore Docker update will come in another PR
- kvstore resiliency tests will come in another PR

Fixes liftedinit/many-framework#241 

- `many-ledger` and `many-kvstore` share a lot of the same `account` & `event` code. We should DRY/refactor this in the future
- Some DRY/TODO in e2e Bash code to be fixed in another PR
- Allowlist whitelisting will come in another PR

Depends on https://github.com/liftedinit/many-rs/pull/203

fmorency committed:
Refactor KvStore

- Add account/event support
- Add query endpoint
- Rename `delete` to `disable`
- Add tests

fmorency committed:
Merge remote-tracking branch 'origin/main' into fmorency/revisit-kvstore

fmorency committed:
.

fmorency committed:
clippy

fmorency committed:
Fix TODO

fmorency committed:
Lots of cleanup and tests

fmorency committed:
oops

fmorency committed:
Install kvstore for bats tests

fmorency committed:
Fix dependencies

fmorency committed:
Remove echo

fmorency committed:
Enable ecdsa and ed25519

fmorency committed:
Add missing endpoint

fmorency committed:
Add kvstore verbosity and fix staging file

fmorency committed:
Make kvstore utility async-aware

fmorency committed:
Cleanup

fmorency committed:
More cleanup

fmorency committed:
Merge remote-tracking branch 'origin/main' into fmorency/revisit-kvstore

hansl commented on src/kvstore/src/main.rs:
```diff
@@ -67,6 +82,16 @@ struct GetOpt {
+    key: String,
+
+    /// If the key is passed as an hexadecimal string, pass this key.
+    #[clap(long)]
```
Should be mutually exclusive.

fmorency replied:
`hex_key` is a flag. If set to `true`, it means `key` is a hex string to decode.

hansl commented on src/kvstore/src/main.rs:
```diff
@@ -85,6 +110,20 @@ struct PutOpt {
+
+    /// If the key is a hexadecimal string, pass this flag.
+    #[clap(long)]
+    hex_key: bool,
```
Should be mutually exclusive with `key`.

fmorency replied:
Same here, `hex_key` is a flag.

hansl commented on src/kvstore/src/main.rs:
```diff
@@ -96,46 +135,155 @@ fn get(client: ManyClient<impl Identity>, key: &[u8], hex: bool) -> Result<(), M
+                    println!("{} ({})", owner, b)
+                }
+                Either::Right(reason) => {
+                    println!("{}, ({})", owner, reason)
```
```suggestion
                    println!("{}, disabled ({})", owner, reason)
```

hansl commented on src/kvstore/src/main.rs:
```diff
@@ -96,46 +135,155 @@ fn get(client: ManyClient<impl Identity>, key: &[u8], hex: bool) -> Result<(), M
+        if let Some(disabled) = result.disabled {
+            match disabled {
+                Either::Left(b) => {
+                    println!("{} ({})", owner, b)
```
```suggestion
                    println!("{}, disabled", owner)
```

fmorency replied:
I'll fix it but your suggestion is wrong since `b` can be `false`.

hansl reviewed (changes requested):
This is a lot of code, I'll continue reviewing tomorrow.

hansl commented on src/many-kvstore/src/module/account.rs:
```diff
@@ -0,0 +1,316 @@
+        if let Some(account) = self.storage.get_account(alternative_owner) {
+            account.needs_role(sender, roles)
+        } else if alternative_owner.is_subresource() {
+            Err(error::subres_alt_unsupported())
```
Why?

hansl commented on src/many-kvstore/src/module/account.rs:
```diff
@@ -0,0 +1,316 @@
+        if let Some(account) = self.storage.get_account(alternative_owner) {
+            account.needs_role(sender, roles)
+        } else if alternative_owner.is_subresource() {
+            Err(error::subres_alt_unsupported())
```
Why?

fmorency replied:
Ah, yes.

I will re-verify, but I think we're missing some bits in `many-rs` to be able to verify that a subresource belongs to a given parent address. I didn't want to switch context and just disallowed it for now.

I can add it in a future PR.

WDYT?

fmorency replied:
https://github.com/liftedinit/many-framework/issues/246

hansl commented on src/many-kvstore/tests/events.rs:
```diff
@@ -0,0 +1,170 @@
+        .unwrap();
+    // TODO: Remove this when we support factional seconds
+    // See https://github.com/liftedinit/many-rs/issues/110
+    std::thread::sleep(std::time::Duration::new(1, 0));
```
You can just do `before + 1`.

hansl commented on src/many-kvstore/tests/events.rs:
```diff
@@ -0,0 +1,170 @@
+
+    // TODO: Remove this when we support factional seconds
+    // See https://github.com/liftedinit/many-rs/issues/110
+    std::thread::sleep(std::time::Duration::new(1, 0));
```
Here too, `let now = after + 1;` should work.

hansl commented on src/many-ledger/src/main.rs:
```diff
@@ -1,7 +1,8 @@
 use many_identity::verifiers::AnonymousVerifier;
 #[cfg(feature = "balance_testing")]
-use many_identity::{Address, Identity};
+use many_identity::Address;
+use many_identity::Identity;
```
```suggestion
use many_identity::{Address, Identity};
```

fmorency replied:
So I merged it based on the suggestion but I remember why I did this; there's a feature guard one line above.

`Address` is only used with `balance_testing` while `Identity` is required all the time.

hansl reviewed (changes requested):
More comments. Tomorrow I'll review the storage and tests.

fmorency committed:
Update Cargo.toml

Co-authored-by: Hans Larsen <hans@larsen.online>
Signed-off-by: Félix C. Morency <1102868+fmorency@users.noreply.github.com>

fmorency committed:
Update src/kvstore/src/main.rs

Co-authored-by: Hans Larsen <hans@larsen.online>
Signed-off-by: Félix C. Morency <1102868+fmorency@users.noreply.github.com>

fmorency committed:
Update src/many-ledger/src/main.rs

Co-authored-by: Hans Larsen <hans@larsen.online>
Signed-off-by: Félix C. Morency <1102868+fmorency@users.noreply.github.com>

fmorency committed:
Address review comments

fmorency committed:
Support binding to another addr

fmorency committed:
Revert "Update src/many-ledger/src/main.rs"

This reverts commit ef450f26d1d18b1fa4e59e17e3bdfceaadcfea82.

fmorency committed:
Add Vergen and signal support to kvstore

fmorency committed:
.

fmorency committed:
Fix blockchain info

fmorency committed:
Resiliency tests

- Refactor Docker e2e to split kvstore and ledger environment

  $ make -f Makefile.common Cargo.nix
  $ make -f Makefile.common many/many-abci
  $ make -f Makefile.ledger many/many-ledger
  $ make -f Makefile.kvstore many/many-kvstore

- Cleanup dependencies
- Add kvstore tests to CI
- Nightly kvstore docker image

fmorency committed:
Fix nightly bats test path

hansl commented:
```bash
# Refactor Docker e2e to split kvstore and ledger environment

  $ make -f Makefile.common Cargo.nix
  $ make -f Makefile.common many/many-abci
  $ make -f Makefile.ledger many/many-ledger
  $ make -f Makefile.kvstore many/many-kvstore
```

My first question would be why not put them in the same Makefile? Is there any value to `make -f Makefile.ledger many/many-ledger` instead of `make many/many-ledger`? There doesn't seem to be duplication of rules in either makefiles.

hansl commented:
Also, is https://www.gnu.org/software/make/manual/html_node/Include.html a viable option?

fmorency commented:
> Also, is https://www.gnu.org/software/make/manual/html_node/Include.html a viable option?

Yes, I'm already using it in `Makefile.ledger` and `Makefile.kvstore`

fmorency commented:
> ```shell
> # Refactor Docker e2e to split kvstore and ledger environment
> 
>   $ make -f Makefile.common Cargo.nix
>   $ make -f Makefile.common many/many-abci
>   $ make -f Makefile.ledger many/many-ledger
>   $ make -f Makefile.kvstore many/many-kvstore
> ```
> 
> My first question would be why not put them in the same Makefile? Is there any value to `make -f Makefile.ledger many/many-ledger` instead of `make many/many-ledger`? There doesn't seem to be duplication of rules in either makefiles.

I can create a standard `Makefile` with our most used commands, which would make everything more dev-friendly, e.g.,

```
$ make Cargo.nix
$ make many/many-abci
$ make many/many-ledger
$ make many/many-kvstore
```

This new `Makefile` would call the others.

The current architecture makes it easy to isolate/switch environments without changing the call, i.e., one can

```
$ make -f Makefile.ledger start-nodes
$ make -f Makefile.kvstore start-nodes
```

It also makes it easy if we ever want to split this mono-repo.

WDYT?

fmorency committed:
Add docker/e2e/genfiles* to .gitignore

hansl commented:
> I can create a standard Makefile with our most used commands, which would make everything more dev-friendly, e.g.,

I think that would be better for devs. Makefile can still be split (like you said will help when we split repo), but at least have a "generic" one. Maybe `start-ledger` and/or `start-kvstore` etc?

hansl reviewed (changes requested):
I can't seem to be able to run the bats tests on my computer:

<img width="2885" alt="image" src="https://user-images.githubusercontent.com/681969/192613457-27b53b69-4f96-4987-a982-bb922e1a3346.png">

Other tests pass.

fmorency commented:
> I can't seem to be able to run the bats tests on my computer:
> 
> Other tests pass.

This is the error we **sometimes** see on the CI, **only** on macos.

Are you able to reproduce systematically? On CI it only happens from time to time, and re-running the failed task usually works.

Do you have some time to investigate?

fmorency committed:
Update src/kvstore/src/main.rs

Co-authored-by: Hans Larsen <hans@larsen.online>
Signed-off-by: Félix C. Morency <1102868+fmorency@users.noreply.github.com>

fmorency commented on src/many-kvstore/src/module/account.rs:
```diff
@@ -0,0 +1,316 @@
+    }
+
+    /// Verify if user is permitted to access the value at the given key
+    pub(crate) fn verify_acl(&self, sender: &Address, key: Vec<u8>) -> Result<(), ManyError> {
```
According to `clippy`, this and all the other `#[must_use]` suggestions return

```
error: this function has an empty `#[must_use]` attribute, but returns a type already marked as `#[must_use]`
```

fmorency commented on src/many-kvstore/src/module/account.rs:
```diff
@@ -0,0 +1,316 @@
+    Ok(())
+}
+
+fn validate_roles_for_account(account: &account::Account) -> Result<(), ManyError> {
```
https://github.com/liftedinit/many-framework/pull/242#discussion_r981697873

fmorency commented on src/many-kvstore/src/module/account.rs:
```diff
@@ -0,0 +1,316 @@
+    Ok(())
+}
+
+fn validate_features_for_account(account: &account::Account) -> Result<(), ManyError> {
```
https://github.com/liftedinit/many-framework/pull/242#discussion_r981697873

fmorency commented on src/many-kvstore/src/module/account.rs:
```diff
@@ -0,0 +1,316 @@
+use std::collections::{BTreeMap, BTreeSet};
+use std::fmt::{Debug, Formatter};
+
+pub(crate) fn validate_account(account: &account::Account) -> Result<(), ManyError> {
```
https://github.com/liftedinit/many-framework/pull/242#discussion_r981697873

fmorency commented on src/many-kvstore/src/module/account.rs:
```diff
@@ -0,0 +1,316 @@
+impl KvStoreModuleImpl {
+    /// Verify the alternative owner is supported
+    /// Verify the sender has the rights to use this alternative owner address
+    pub(crate) fn validate_alternative_owner<R: TryInto<Role> + std::fmt::Display + Copy>(
```
https://github.com/liftedinit/many-framework/pull/242#discussion_r981697873

fmorency committed:
Fix from review

hansl commented:
> This is the error we sometimes see on the CI, only on macos.

I can take a look.

hansl commented:
Good work.

Denommus reviewed (approved):
lgtm

fmorency committed:
Update to latest `many-rs` & simplify CI
